### PR TITLE
Serialize units ratelimits and document StartLimitIntervalSec=infinity

### LINKED ISSUE #29574: StartLimitIntervalSec=infinity is undocumented, and also is not preserved across reloads
### systemd version the issue has been seen with

253

### Used distribution

Fedora 38

### Linux kernel version used

6.5.6-200.fc38.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemctl

### Expected behaviour you didn't see

Manual of systemd.unit(5) doesn't state that you can use `infinity` as an interval for `StartLimitIntervalSec=` option, however, it seems to work as intuition implies (allows `StartLimitBurst` value to be contained within an unlimited timeframe).

### Unexpected behaviour you saw

`infinity` works despite being undocumented

### Steps to reproduce the problem

create a systemd unit with `StartLimitBurst=5` and `StartLimitIntervalSec=infinity` options, make it so that it fails on start (like `ExecStart=/bin/false`), then start it and observe the journal. It will stop restarting after restart count reaches 5 no matter the `RestartSec=` value.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by keszybz**
This seems to work as expected. In the issue, doubts were raised whether it works fine with daemon-reload/daemon-reexec, and it seems to work fine. (The property cannot be set via set-property, the dbus property is 'const'. We could relax this, but that'd be a separate feature.)

Closes #29574.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

**Comment by Werkov** [comment]
> daemon-reload/daemon-reexec, and it seems to work fine.

I meant the internal state of `start_ratelimit` doesn't seem to be serialized.
The effect is pronounced with infinity timespan
(i.e. you allow finite number of failures but it gets implicitly reset upon those reloads).

**Comment by keszybz** [comment]
Oh, OK. Yeah, if the interval is small, this doesn't really matter. But with an infinite interval, it makes a difference. I'll try to fix this.

File: src/shared/serialize.c [hunk 1]
```diff
@@ -201,6 +201,15 @@ int serialize_pidref(FILE *f, FDSet *fds, const char *key, PidRef *pidref) {
         return serialize_item_format(f, key, "@%i", copy);
 }
 
+int serialize_ratelimit(FILE *f, const char *key, const RateLimit *rl) {
+        return serialize_item_format(f, key,
```
Comment (bluca) [inline src/shared/serialize.c:207]: assert(rl)

File: src/shared/serialize.c [hunk 2]
```diff
@@ -486,6 +495,18 @@ int deserialize_pidref(FDSet *fds, const char *value, PidRef *ret) {
         return 0;
 }
 
+void deserialize_ratelimit(RateLimit *rl, const char *name, const char *value) {
+        usec_t begin, interval;
+        unsigned num, burst;
+
```
Comment (bluca) [inline src/shared/serialize.c:503]: assert(rl)

File: src/core/unit-serialize.c [hunk 3]
```diff
@@ -140,6 +140,8 @@ int unit_serialize_state(Unit *u, FILE *f, FDSet *fds, bool switching_root) {
         (void) serialize_dual_timestamp(f, "condition-timestamp", &u->condition_timestamp);
         (void) serialize_dual_timestamp(f, "assert-timestamp", &u->assert_timestamp);
 
+        (void) serialize_ratelimit(f, "start-ratelimit", &u->start_ratelimit);
```
Comment (poettering) [inline src/core/unit-serialize.c:143]: hmm, the cherry on top would be if you'd also add this for `u->auto_start_stop_ratelimit`.

And then also for the `trigger_limit` of Path and Socket objects

**Comment by keszybz** [comment]
Updated to serialize/deserialize trigger ratelimits and auto start/stop ratelimits.

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
Both fails in the Arch Linux jobs are unrelated (though concerning, as I've seen them in multiple CI runs already).

**Comment by mrc0mmand** [comment]
Ah, nevermind, the fails _are_ related, I just didn't refresh the page so it looked like a man-only change.

File: src/core/manager.c [hunk 4]
```diff
@@ -927,10 +927,7 @@ int manager_new(RuntimeScope runtime_scope, ManagerTestRunFlags test_run_flags,
                 .first_boot = -1,
                 .test_run_flags = test_run_flags,
 
-                .dump_ratelimit = {
-                        .interval = 10 * USEC_PER_MINUTE,
-                        .burst = 10,
-                },
+                .dump_ratelimit = { .interval = 10 * USEC_PER_MINUTE, .burst = 10 },
```
Comment (poettering) [inline src/core/manager.c]: here you specify the field names

File: src/core/unit.c [hunk 5]
```diff
@@ -134,13 +134,10 @@ Unit* unit_new(Manager *m, size_t size) {
...
                 m->defaults.start_limit_interval,
-                m->defaults.start_limit_burst
+                m->defaults.start_limit_burst,
         };
 
-        u->auto_start_stop_ratelimit = (const RateLimit) {
-                10 * USEC_PER_SEC,
-                16
-        };
+        u->auto_start_stop_ratelimit = (const RateLimit) { 10 * USEC_PER_SEC, 16 };
```
Comment (poettering) [inline src/core/unit.c]: here you don't

**Comment by poettering** [comment]
lgtm

**Comment by mrc0mmand** [comment]
Ah, since the rate limits now persist through daemon-reload, something like:

```diff
diff --git a/test/units/testsuite-04.SYSTEMD_JOURNAL_COMPRESS.sh b/test/units/testsuite-04.SYSTEMD_JOURNAL_COMPRESS.sh
index 8fdf1b132e..96d096d9ad 100755
--- a/test/units/testsuite-04.SYSTEMD_JOURNAL_COMPRESS.sh
+++ b/test/units/testsuite-04.SYSTEMD_JOURNAL_COMPRESS.sh
@@ -7,6 +7,9 @@ set -o pipefail
 mkdir /run/systemd/system/systemd-journald.service.d
 MACHINE_ID="$(</etc/machine-id)"
 
+# Reset the start-limit counters, as we're going to restart journald a couple of times
+systemctl reset-failed systemd-journald.service
+
 for c in NONE XZ LZ4 ZSTD; do
     cat >/run/systemd/system/systemd-journald.service.d/compress.conf <<EOF
 [Service]
@@ -35,4 +38,5 @@ done
 rm /run/systemd/system/systemd-journald.service.d/compress.conf
 systemctl daemon-reload
 systemctl restart systemd-journald.service
+systemctl reset-failed systemd-journald.service
 journalctl --rotate
```

should make the CIs happy again.

File: src/core/manager.c [hunk 6]
```diff
@@ -960,7 +957,7 @@ int manager_new(RuntimeScope runtime_scope, ManagerTestRunFlags test_run_flags,
         }
 
         /* Reboot immediately if the user hits C-A-D more often than 7x per 2s */
-        m->ctrl_alt_del_ratelimit = (const RateLimit) { .interval = 2 * USEC_PER_SEC, .burst = 7 };
+        m->ctrl_alt_del_ratelimit = (RateLimit) { .interval = 2 * USEC_PER_SEC, .burst = 7 };
```
Comment (poettering) [inline src/core/manager.c]: why drop the const?

File: src/core/socket.c [hunk 7]
```diff
@@ -100,8 +100,7 @@ static void socket_init(Unit *u) {
         s->control_pid = PIDREF_NULL;
         s->control_command_id = _SOCKET_EXEC_COMMAND_INVALID;
 
-        s->trigger_limit.interval = USEC_INFINITY;
-        s->trigger_limit.burst = UINT_MAX;
+        s->trigger_limit = (RateLimit) { USEC_INFINITY, UINT_MAX };
```
Comment (poettering) [inline src/core/socket.c]: here you don't specify the field names.

maybe we should define RATELIMIT_OFF btw, whith this definition

File: src/core/manager.c (hunk 6 above)
Comment (yuwata) [reply src/core/manager.c]: Why `const` is necessary?

**Comment by bluca** [comment]
How are we doing here? Is this good to merge?

File: src/core/manager.c (hunk 6 above)
Comment (YHNdnzj) [reply src/core/manager.c]: The same reason as https://github.com/systemd/systemd/commit/9bb64f1b8d2252d558f209a04b171d65b5ef131b I reckon.

File: src/core/unit-serialize.c [hunk 8]
```diff
@@ -343,6 +346,13 @@ int unit_deserialize_state(Unit *u, FILE *f, FDSet *fds) {
                         (void) deserialize_dual_timestamp(v, &u->assert_timestamp);
                         continue;
 
+                } else if (streq(l, "start-ratelimit")) {
+                        deserialize_ratelimit(&u->start_ratelimit, "start-ratelimit", v);
```
Comment (YHNdnzj) [inline src/core/unit-serialize.c]: Maybe simply pass `l` instead?

File: src/core/unit-serialize.c [hunk 9]
```diff
@@ -343,6 +346,13 @@ int unit_deserialize_state(Unit *u, FILE *f, FDSet *fds) {
...
+                        continue;
+                } else if (streq(l, "auto-start-stop-ratelimit")) {
+                        deserialize_ratelimit(&u->auto_start_stop_ratelimit, "auto-start-stop-ratelimit", v);
```
Comment (YHNdnzj) [inline src/core/unit-serialize.c]: also here

File: src/core/socket.c [hunk 10]
```diff
@@ -2824,7 +2825,10 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
                 if (!found)
                         log_unit_debug(u, "No matching ffs socket found: %s", value);
 
-        } else
+        } else if (streq(key, "trigger-ratelimit"))
+                deserialize_ratelimit(&s->trigger_limit, "trigger-ratelimit", value);
```
Comment (YHNdnzj) [inline src/core/socket.c]: and here

File: src/core/path.c [hunk 11]
```diff
@@ -744,7 +745,10 @@ static int path_deserialize_item(Unit *u, const char *key, const char *value, FD
                                 }
                 }
 
-        } else
+        } else if (streq(key, "trigger-ratelimit"))
+                deserialize_ratelimit(&p->trigger_limit, "trigger-ratelimit", value);
```
Comment (YHNdnzj) [inline src/core/path.c]: and here

**Review by YHNdnzj — APPROVED** [review]

**Pull request merged**